Feature: add data collector package (#127)
project-march/tutorials: doc/march_packages/march_data_collector.rst

2.
    Create a callback function that takes a ROS message and calls the ``send_to_esp`` function with the msgs as csv
    string in :march:`esp_adapter.py <march_data_collector/src/march_data_collector/esp_adapter.py>`.

3.
    In the same file add a call to ``configure_source`` to the ``__init__``.

Pressure Soles
--------------
The `march_data_collector` interfaces between the Moticon pressure soles and the exoskeleton software.
It  sends data to and receives data from the desktop software that came with the pressure soles.
This data is send over `UDP <https://nl.wikipedia.org/wiki/User_Datagram_Protocol>`_.
More information on interacting with the Moticon software can be found `here <https://www.moticon.de/doc/science_desktop_software/record/udp/>`_.
The package assumes that the filter used in the Moticon software is ``force pressure cop``.
More information on using the pressure soles can be found at :ref:`using the pressure soles <using-the-pressure-soles-label>`.

ROS API
-------

Nodes
^^^^^
*march_data_collector_node* - Responsible for processing IMU data and calculating the acceleration and jerk.

*com_calculator* - Responsible for calculating the center of mass.

> *cp_calculator* - Responsible for calculating the capture points.

Reviewer: These are technically not separate ROS nodes, but are included in the `march_data_collector_node`.

*esp_adapter* - Responsible for sending data to the ESP engine.

Subscribed Topics
^^^^^^^^^^^^^^^^^
*/march/controller/trajectory/follow_joint_trajectory/feedback* (control_msgs/JointTrajectoryControllerState)
  Compute the acceleration and jerk.

*/march/imu* (sensor_msgs/Imu)
  Use IMU data to change the orientation of the tf frames and send to `ESP`.

*/march/pressure_soles* (march_shared_resources/PressureSole)
  Send pressure sole data to `ESP`.

*/march/imc_states* (march_shared_resources/ImcState)
  Send iMOTIONCUBE data to `ESP`.

*/march/gait/schedule/goal* (march_shared_resources/GaitActionGoal)
  Send gait data to `ESP`.

*/march/com_marker* (visualization_msgs/Marker)
  Send center of mass to `ESP`.

*/march/joint_values* (march_shared_resources/JointValues)
  Send position, velocity, acceleration and jerk of joints to `ESP`.

*/march/joint_states* (sensor_msgs/JointState)
  Send actual effort to `ESP`.

> */march/controller/after_limit_joint_command'* (march_shared_resources/AfterLimitJointCommand)

Reviewer: ```suggestion
*/march/controller/after_limit_joint_command* (march_shared_resources/AfterLimitJointCommand)
```

.. _march-data-collector-label:

march_data_collector
====================

Overview
--------
The `march_data_collector` package subscribes to the topics on which data is published to process the data. Basically three things are happening to the data:

> * Based on the position, the acceleration and jerk are calculated for future analysis.

Reviewer 2: ```suggestion
* Based on the position, the acceleration and jerk (through second and third order backward Euler) are calculated for future analysis.
```

* The IMU is used to adapt the tf frames and based on this the center of mass (CoM) and the capture point (CP) are calculated.

* The data is send to an `Event Stream Processing` Engine.

> Furthermore the  package also publishes incoming data from the pressure soles and sends data to the software coming with the pressure soles.

Reviewer 2: ```suggestion
Furthermore the  package also publishes incoming data from the pressure soles and sends data to the software that came with the pressure soles.
```

Calculating `CoM` and `CP`
--------------------------
The data from the IMU (:ref:`march-imu-manager-label`) is used to change the tf frames (see :ref:`robot-model-label`).
such that the tf frames are orientated with respect to the world correctly.
The tf frames together with the masses of the links from the :ref:`URDF <robot-model-label>` can be used to calculate the center of mass.
This is then published as a marker on ``/march/com_marker``.
The calculation is based on an implementation from `Hamburg Bit-Bots <https://github.com/bit-bots>`_.

> The marker can also be visualized in :ref:`rviz <robot-model-label>`. The tf frames and CoM together can be used to calculate the capture point.

Reviewer 2: ```suggestion
The marker can also be visualized in :ref:`rviz <robot-model-label>`. The tf frames and CoM can be combined to calculate the capture point.
```

> This is the point in a walking gait where one should step in order to return to a stable standing position.

Reviewer 2: ```suggestion
This is the point where one should place their foot in order to return to a stable standing position.
```

This calculation is done for both feet and published as a marker on ``/march/cp_marker_ankle_plate_{left|right}``
The center of mass and capture point are used for (research on) balancing the exoskeleton.

Event Stream Processing
-----------------------
An `Event Stream Processing` engine from `SAS <https://www.sas.com/nl_nl/home.html>`_ is used.
There is a lot of documentation on this engine at
`ESP documentation <https://documentation.sas.com/?cdcId=espcdc&cdcVersion=6.2&docsetId=espov&docsetTarget=home.htm&locale=nl>`_.

> The Engine allows for real-time data analysis and processing.

Reviewer 2: ```suggestion
The engine allows for real-time data analysis and processing.
```

An engine consists of multiple windows. The data comes in at so-called `source windows`.
From the source windows data is send to other windows with different functions, like joining different event streams, aggregating an
event stream, transforming an event stream or using the data in the events for an algorithm. The windows should form an acyclic graph together.
A group of windows is called a :march:`model <march_data_collector/esp_models/march.xml>`.

> Other applications can  subscribe to a windows to use the data in the window. This can be done for data visualisation, csv output or streaming back into our software.

Reviewer 2: ```suggestion
Other applications can subscribe to a windows to use the data in the window. This can be done for data visualisation, csv output or streaming back into our software.
```
When you pay too much attention to the details :sweat_smile:

The :march:`esp_adapter.py <march_data_collector/src/march_data_collector/esp_adapter.py>` subscribes to ROS msgs and transforms incoming messages to csv strings. The csv strings are send to the ESP engine.

> Pressure Soles

Reviewer 2: Should this be inside te march_data_collector package documentation? It does not really have much to do with the code.

Author: Where would you put it? As a seperate tutorial under ``using the March exoskeleton``?

Reviewer 2: Yes, I think that is the right place. You do not really need to know anything about the code to use the pressure soles, and you do not need to know precisely how to use the pressure soles to edit the code. So I think usage of data collectors and documentation of the code should be split.

1. First you open the app on the phone and pair the insoles with the phone.

.. note::

    The pressure soles are turned on by moving them, because the internal IMU is triggered.
    Shake the pressure soles when they do not automatically turn on.

2. Next you should set the record mode to live capture. Open the Moticon desktop app and go to the record section and settings. Set the following things:

    * UDP Input:
        - `Channel name:`           Joint_positions
        - `Number rof Channels:`    8
        - `Datatype:`               float32
        - `Port:`                   9999

    * UDP Output
        - `IP addess:`              \<exoskeleton_ip\>
        - `Port:`                   8888
        - `Filter:`                 force pressure cop

More info on this step can be found `here <https://www.moticon.de/doc/science_desktop_software/record/udp/>`_.

3. Lastly use should set the correct launch arguments. You should set pressure_soles to true and if you are not using the
standard Switch laptop connected to the project March router, you should set the IP-address on which the Moticon desktop app is running.

ROS API
-------

Nodes
^^^^^

> *march_data_collector_node* - Responsible for processing IMU data and calculating the jerk.

Reviewer 2: ```suggestion
*march_data_collector_node* - Responsible for processing IMU data and calculating the acceleration and jerk.
```

  Send effort command to `ESP`.

*/march/temperature/\** (sensor_msgs/Temperature)
  Send temperature command to `ESP` for each joint.

Subscribes to the tf frames trough a `tf2 <http://wiki.ros.org/tf2>`_ buffer.

Published Topics
^^^^^^^^^^^^^^^^
*/march/com_marker* (visualization_msgs/Marker)
  A marker with the CoM point.

*/march/cp_marker_ankle_plate_left* (visualization_msgs/Marker)
  A marker with the capture point for the left foot.

*/march/cp_marker_ankle_plate_right* (visualization_msgs/Marker)
  A marker with the capture point for the right foot.

*/march/pressure_soles* (march_shared_resources/PressureSole)
  A marker with the capture point for the right foot.

*/march/pressure_soles* (march_shared_resources/JointValues)
  Send the data from the pressure soles.

Publishes into tf via a `tf2 <http://wiki.ros.org/tf2>`_ broadcaster.

Parameters
^^^^^^^^^^
*/march/march_data_collector/moticon_ip* (*string*, default: 192.168.8.105)
  IP-address the Moticon desktop software is running
*/march/march_data_collector/pressure_soles* (*bool*, default: false)
  Whether to connect with the pressure soles.

> Tutorials

Reviewer 2: I think you should add an introduction to explain what ESP does. It listens to ros messages and outputs them to a csv file right? It think it would be nice to explain this and give an example of what you can use it for.

Author: Would you make a seperate tuturial on it and reference that one here and only mention the integration here?

Reviewer 2: O I see now that there is already an explanation in the "Event Stream Processing" section. Perhaps it is possible to move the Esp tutorials to the ESP chapter? I think such a structure would be a bit easier to follow. 
Also, I still do not really understand how one would use ESP. I get now how you subscribe to topics and how you create windows, but how do you use the output? How do you finally make the plots you want/get the mean squared error etc. Perhaps it is possible to add an example with a typical usage?
Perhaps it becomes too long then and you should move the non-code part to a seperate page, perhaps the same page as the pressure soles.
